rust/src/operations/transaction.rs

```diff
@@ -0,0 +1,193 @@
+            .iter()
+            .map(serde_json::to_string)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(to_datafusion_err)?,
+    );
+    Ok(RecordBatch::try_new(
+        Arc::new(OPERATION_SCHEMA.clone()),
+        vec![Arc::new(serialized)],
+    )?)
+}
+
+/// Write command
+#[derive(Debug)]
+pub struct DeltaTransactionPlan {
+    table_uri: String,
+    table_version: DeltaDataTypeVersion,
+    input: Arc<dyn ExecutionPlan>,
+    operation: DeltaOperation,
+    app_metadata: Option<serde_json::Map<String, serde_json::Value>>,
+}
+
+impl DeltaTransactionPlan {
+    /// Wrap partitioned delta operations in a DeltaTransaction
+    pub fn new<T>(
+        table_uri: T,
+        table_version: DeltaDataTypeVersion,
+        input: Arc<dyn ExecutionPlan>,
+        operation: DeltaOperation,
+        app_metadata: Option<serde_json::Map<String, serde_json::Value>>,
+    ) -> Self
+    where
+        T: Into<String>,
+    {
+        Self {
+            table_uri: table_uri.into(),
+            table_version,
+            input: Arc::new(CoalescePartitionsExec::new(input)),
+            operation,
+            app_metadata,
+        }
+    }
+
+    /// Arrow schema expected to be produced by wrapped operations
+    pub fn input_operation_schema() -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+}
+
+#[async_trait]
+impl ExecutionPlan for DeltaTransactionPlan {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn required_child_distribution(&self) -> Distribution {
+        Distribution::UnspecifiedDistribution
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        todo!()
+    }
+
+    async fn execute(
+        &self,
+        _partition: usize,
+        context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        let mut table =
+            get_table_from_uri_without_update(self.table_uri.clone()).map_err(to_datafusion_err)?;
+
+        let data = collect(self.input.clone(), context.clone()).await?;
+        // TODO we assume that all children send a single column record batch with serialized actions
+        let actions = data
+            .iter()
+            .flat_map(|batch| match deserialize_actions(batch) {
+                Ok(vec) => vec.into_iter().map(Ok).collect(),
+                Err(er) => vec![Err(er)],
+            })
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
I think if you separate `map` and `flatten`, it will populate the error nicely:

```suggestion
            .map(deserialize_actions)
            .flatten()
            .collect::<Result<Vec<_>, _>>()?;
```

Reply:
turns out this would need to chained `flattens`, definately a matter of taste, but in that case i personally like the explicit matching better :).